### PR TITLE
Copying between memory and memory.

### PR DESCRIPTION
**Pull request by chriseth**
Also fixed some encoding and padding issues with older copying code.

### CODE REVIEW

File: libsolidity/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -550,7 +436,7 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 			_functionCall.getExpression().accept(*this);
 			m_context << u256(0); // do not send gas (there still is the stipend)
 			arguments.front()->accept(*this);
-			appendTypeConversion(*arguments.front()->getType(),
+			utils().convertType(*arguments.front()->getType(),
```
Comment (LianaHus) [inline libsolidity/ExpressionCompiler.cpp]: intend

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 2]
```diff
@@ -4232,6 +4232,31 @@ BOOST_AUTO_TEST_CASE(reusing_memory)
...
+//			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "Main");
+	string s("Julia");
+	bytes args = encodeArgs(u256(0x20), u256(s.length()), s);
+	BOOST_REQUIRE(callContractFunction("set(string)", asString(args)) == encodeArgs());
+	BOOST_CHECK(callContractFunction("get1()") == args);
+//	BOOST_CHECK(callContractFunction("get2()") == args);
+//	BOOST_CHECK(callContractFunction("s()") == args);
```
Comment (LianaHus) [inline test/libsolidity/SolidityEndToEndTest.cpp]: do you really  whant to leave those comments?

File: libsolidity/ExpressionCompiler.cpp [hunk 3]
```diff
@@ -726,12 +612,12 @@ void ExpressionCompiler::endVisit(MemberAccess const& _memberAccess)
...
 		{
-			appendTypeConversion(*_memberAccess.getExpression().getType(),
+			utils().convertType(*_memberAccess.getExpression().getType(),
 								 IntegerType(0, IntegerType::Modifier::Address), true);
 			m_context << eth::Instruction::BALANCE;
 		}
 		else if ((set<string>{"send", "call", "callcode"}).count(member))
-			appendTypeConversion(*_memberAccess.getExpression().getType(),
+			utils().convertType(*_memberAccess.getExpression().getType(),
 								 IntegerType(0, IntegerType::Modifier::Address), true);
```
Comment (LianaHus) [inline libsolidity/ExpressionCompiler.cpp]: intend, same at the line 616

File: libsolidity/CompilerUtils.cpp [hunk 4]
```diff
@@ -159,6 +237,288 @@ void CompilerUtils::storeInMemoryDynamic(Type const& _type, bool _padToWordBound
...
+			// clear lower-order bytes for conversion to shorter bytes - we always clean
+			solAssert(targetTypeCategory == Type::Category::FixedBytes, "Invalid type conversion requested.");
+			FixedBytesType const& targetType = dynamic_cast<FixedBytesType const&>(_targetType);
+			if (targetType.getNumBytes() < typeOnStack.getNumBytes())
+			{
+				if (targetType.getNumBytes() == 0)
+					m_context << eth::Instruction::DUP1 << eth::Instruction::XOR;
+				else
+					m_context << (u256(1) << (256 - targetType.getNumBytes() * 8))
+							  << eth::Instruction::DUP1 << eth::Instruction::SWAP2
```
Comment (LianaHus) [inline libsolidity/CompilerUtils.cpp]: indent

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 2 above)
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: They will be removed in the next pull request, they still fail here.

**Pull request merged**